### PR TITLE
[global-hooks] Fix error on change kubernetesVersion to Automatic

### LINKED ISSUE #5497: [control-plane-manager] Changing kubernetesVersion to Automatic can trigger downgrade by more than 1 version
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Version

v1.49.3

### Expected Behavior

Deckhouse prevents downgrade that can break the cluster.

### Actual Behavior

If I use `kubernetesVersion`, for example 1.27, and then change it to `Automatic`, deckhouse will start downgrading to the default version (currently 1.23)

### Steps To Reproduce

_No response_

### Additional Information

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by nikiGGG**
## Description

Union cluster configuration and kubernetes version hooks.
Add validation webhook for d8-cluster-configuration secret for prevent downgrade k8s version >= 2 minor versions.
Add check for prevent downgrade k8s version >= 2 minor versions in cluster_configuration global hook when user change k8s version to automatic 

## Why do we need it, and what problem does it solve?

Close #5497 and #6855

## What is the expected result?

initial state
image

can downgrade on one version
image
image

cannot downgrade to -2 minor version from maximum used in cluster
image

Cannot downgrade to Automatic
image

Allow upgrade
image
image

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: global-hooks
type: fix
summary: Cannot downgrade kubernetes version more than 1 version.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @nikiGGG. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @nikiGGG. I've started the workflow for label 'e2e/run/openstack'...

**Comment by nikiGGG** [comment]
/e2e/abort openstack;Standard;containerd;1.25 7739352511 7739352511-1-con-1-25 /sys/deckhouse-oss/install:pr7279 debian@188.124.39.98

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @nikiGGG. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @nikiGGG. I've started the workflow for label 'e2e/run/openstack'...

**Comment by nikiGGG** [comment]
Tested on cluster on 1.27 rev version. max kube version is 1.28. Default kubernetes version is 1.25
Снимок экрана 2024-02-04 в 17 49 04
Error when trying to change to 1.25:
Снимок экрана 2024-02-04 в 17 51 53
Succesfully change to 1.28 when set to Automatic:
Снимок экрана 2024-02-04 в 17 53 25

File: global-hooks/discovery/cluster_configuration.go [hunk 1]
```diff
@@ -91,8 +97,10 @@ func clusterConfiguration(input *go_hook.HookInput) error {
 		}
 
 		if kubernetesVersionFromMetaConfig == "Automatic" {
-			b, _ := json.Marshal(config.DefaultKubernetesVersion)
-			metaConfig.ClusterConfig["kubernetesVersion"] = b
+			metaConfig.ClusterConfig["kubernetesVersion"], err = newAutomaticVersion(configYamlBytes.MaxUsedControlPlaneKubernetesVersion, config.DefaultKubernetesVersion)
+		}
+		if err != nil {
```
Comment (aleksey-su) [inline global-hooks/discovery/cluster_configuration.go:102]: The wrong place to check for errors

File: global-hooks/discovery/cluster_configuration.go [hunk 2]
```diff
@@ -168,3 +176,23 @@ func rawMessageToString(message json.RawMessage) (string, error) {
...
+	maxUsedKubernetesVersion, err := semver.NewVersion(maxUsedControlPlaneKubernetesVersion)
+	// Cant parse max used kubenetes version or maxUsedKubernetesVersion equal nil
+	if err != nil || maxUsedKubernetesVersion == nil {
+		return json.Marshal(configDefaultKubernetesVersion)
+	}
+	finalKubernetesVersion := defaultKubernetesVersion
+
+	// Default version will downgrdae kluster more than 1 version
+	// kubernetes version cannot be downgraded more than 1 version: maxUsedKubeVersion=%s, newKubeVersion=%s
+	if (maxUsedKubernetesVersion.Minor() - defaultKubernetesVersion.Minor()) > 1 {
```
Comment (aleksey-su) [inline global-hooks/discovery/cluster_configuration.go:194]: It is necessary to take into account the change in the major version

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration [hunk 3]
```diff
@@ -88,6 +88,17 @@ function __main__() {
       fi
     fi
 
+    # Not allow to downgrade more than 1 kube version
+    newKubernetesVersion=$(yq eval '.kubernetesVersion' - <<< "$newConfig")
+    # Possible kubernetes versions are "Automatic" or semver(1.28) versions
+    if [ "${newKubernetesVersion}" != "Automatic" ]; then
+      maxUsedControlPlaneKubernetesVersion=$(context::jq -r '.review.request.object.data."maxUsedControlPlaneKubernetesVersion"' | base64 -d)
+      if [ $((${maxUsedControlPlaneKubernetesVersion:2} - ${newKubernetesVersion:2})) -gt 1 ]; then
```
Comment (aleksey-su) [inline modules/002-deckhouse/webhooks/validating/d8-cluster-configuration:96]: It is necessary to take into account the change in the major version

**Review by aleksey-su — CHANGES_REQUESTED** [review]

File: modules/002-deckhouse/webhooks/validating/d8-cluster-configuration (hunk 3 above)
Comment (nikiGGG) [reply modules/002-deckhouse/webhooks/validating/d8-cluster-configuration]: check minor and major version. I get it using cut command. After it check all same as in .go file

File: candi/openapi/doc-ru-cluster_configuration.yaml [hunk 4]
```diff
@@ -60,7 +60,8 @@ apiVersions:
 
           Изменение параметра в работающем кластере приведет к [автоматическому обновлению](https://deckhouse.ru/documentation/v1/modules/040-control-plane-manager/#управление-версиями) версии control plane кластера.
 
-          Если указано `Automatic`, то используется версия control plane, которая считается стабильной на данный момент. Изменение этой версии может произойти при смене минорной версии релиза Deckhouse (см. список изменений к релизу).
+          Если указано `Automatic`, то используется версия control plane, которая считается стабильной на данный момент. Но не меньше одной версии от максимальной версии которая когда либо была использована в кластере.
+          Изменение этой версии может произойти при смене минорной версии релиза Deckhouse (см. список изменений к релизу).
```
Comment (z9r5) [inline candi/openapi/doc-ru-cluster_configuration.yaml:64]:
```suggestion
          Если указано `Automatic`, то используется версия control plane, которая считается стабильной на данный момент. Выбираемая автоматически стабильная версия также не может отличаться больше чем на одну минорную версию от максимальной версии, которая когда-либо была использована в кластере.
          Изменение этой версии может произойти, например, при смене минорной версии релиза Deckhouse (см. список изменений к релизу).
```
@MariaMakeeva, your turn...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Comment by name212** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 8176227434 8176227434-1-con-1-25 /sys/deckhouse-oss/install:pr7279

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Review by aleksey-su — APPROVED** [review]

**Comment by z9r5** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 8176227434 8176227434-1-con-1-25 /sys/deckhouse-oss/install:pr7279

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

File: candi/openapi/doc-ru-cluster_configuration.yaml [hunk 5]
```diff
@@ -63,7 +63,8 @@ apiVersions:
 
           Изменение параметра в работающем кластере приведет к [автоматическому обновлению](https://deckhouse.ru/documentation/v1/modules/040-control-plane-manager/#управление-версиями) версии control plane кластера.
 
-          Если указано `Automatic`, то используется версия control plane, которая считается стабильной на данный момент. Изменение этой версии может произойти при смене минорной версии релиза Deckhouse (см. список изменений к релизу).
+          Если указано `Automatic`, то используется версия control plane, которая считается стабильной на данный момент. Если стабильная версия control plane меньше максимальной, которая когда либо была установлена в кластере, больше чем на 1 минорную версию - тогда версия кластера не изменится.
```
Comment (MariaMakeeva) [inline candi/openapi/doc-ru-cluster_configuration.yaml:66]:
```suggestion
          Если указано `Automatic`, то используется та версия control plane, которая считается стабильной на данный момент. Если стабильная версия Control Plane меньше или равна максимальной версии, которая когда-либо была установлена в кластере, и разница составляет более чем 1 минорный выпуск, то версия кластера не изменится.
```

Comment (z9r5) [reply candi/openapi/doc-ru-cluster_configuration.yaml:66]: Здесь не корректно добавлять "или равна ", т.е. если равна, то не может быть разницы версий. Ну и очевидно, что если стабильная  версия control plane равна той, котоаря в кластере - то ничего не изменится.
Тут речь именно про те случаи, когда мы установили в DKP стабильную, например 1.25, а в кластере в какой-то момент была 1.27 (т.е. разница в две минорные версии ), то автоматического изменения версии не произойдет. @MariaMakeeva

**Review by RomanenkoDenys — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Pull request merged**